{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0df73b67-7e12-4d82-8f01-45ba4d8e5b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f384a5df-cc3b-4029-9848-c9522b3aee68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Name', 'Ab or Nb', 'Binds to', 'Doesn't Bind to', 'Neutralising Vs',\n",
      "       'Not Neutralising Vs', 'Protein + Epitope', 'Origin', 'VHorVHH', 'VL',\n",
      "       'Heavy V Gene', 'Heavy J Gene', 'Light V Gene', 'Light J Gene', 'CDRH3',\n",
      "       'CDRL3', 'Structures', 'ABB Homology Model (if no structure)',\n",
      "       'Sources', 'Date Added', 'Last Updated', 'Update Description',\n",
      "       'Notes/Following Up?'],\n",
      "      dtype='object')\n",
      "['S; RBD/non-RBD' 'S; non-RBD' 'S; iso-RBD' 'S; RBD' 'S; S1'\n",
      " 'S; S2 Fusion Peptide' 'S; Unk' 'S; S2 (HR1 Peptide)' 'S; NTD' 'S; S2'\n",
      " 'S; S2 (HR2 Peptide)' 'S; S2 Stem Helix' 'S; NTD-SD1' 'Nsp9'\n",
      " \"S; S2' Cleavage Site/Fusion Peptide NTD\" 'S; SD1' 'S; S1 non-RBD'\n",
      " 'S; Stem Helix' 'N; NTD' 'M' 'S; probably RBD (implied by clustering)'\n",
      " 'CSP ' nan 'Unknown' 'S; S2 (quaternary glycan epitope)'\n",
      " 'S; S1/S2 Cleavage Site' 'N' 'S' 'S; S1/S2' 'E' 'S; Possibly RBD' 'Mpro'\n",
      " 'Nsp3; PLpro' 'N; RBD' 'N; CTD' 'S; non-S1' 'TBC'\n",
      " 'S2 (quaternary glycan epitope)' 'S: NTD' 'S: RBD'\n",
      " \"S; S2' Cleavage Site/Fusion Peptide\" 'S1; non-RBD' 'S; RBD/NTD']\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('CoV-AbDab_080224.csv')\n",
    "print(df.columns)\n",
    "print(df['Protein + Epitope'].unique())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "faf9cdd2-3ac3-4dc4-8514-67411ec5ba80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Ab or Nb</th>\n",
       "      <th>Binds to</th>\n",
       "      <th>Doesn't Bind to</th>\n",
       "      <th>Neutralising Vs</th>\n",
       "      <th>Not Neutralising Vs</th>\n",
       "      <th>Protein + Epitope</th>\n",
       "      <th>Origin</th>\n",
       "      <th>VHorVHH</th>\n",
       "      <th>VL</th>\n",
       "      <th>...</th>\n",
       "      <th>Light J Gene</th>\n",
       "      <th>CDRH3</th>\n",
       "      <th>CDRL3</th>\n",
       "      <th>Structures</th>\n",
       "      <th>ABB Homology Model (if no structure)</th>\n",
       "      <th>Sources</th>\n",
       "      <th>Date Added</th>\n",
       "      <th>Last Updated</th>\n",
       "      <th>Update Description</th>\n",
       "      <th>Notes/Following Up?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8-D9</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2 Human Vaccinee (BBIBP-CorV)</td>\n",
       "      <td>VQLVESGGGLVQPGGSLRLSCAASGLTVSSNYMNWVRQAPGKGLEW...</td>\n",
       "      <td>DIQMTQSPSFLSASVGDRVTITCRASQGISSYLAWYQQKPGKAPKL...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ3 (Human)</td>\n",
       "      <td>ARDHSGHALDI</td>\n",
       "      <td>QHLNSYPSMYT</td>\n",
       "      <td>https://www.rcsb.org/structure/8J1T;https://ww...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wanbo Tai et al., 2023 (https://www.nature.com...</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sun_1G11</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...</td>\n",
       "      <td>SARS-CoV2_Omicron-BQ1;SARS-CoV2_Omicron-BQ1.1;...</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2 Human Patient</td>\n",
       "      <td>EVQLVESGGGLVQPGRSLRLSCAASGFKFDDYAMHWVRQAPGKGLE...</td>\n",
       "      <td>DIQLTQSPSFLSASVGDRVTITCRASQGIDKYLAWYQQKPGQAPKV...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ3 (Human)</td>\n",
       "      <td>VKDSNYDSSGYLINNFDY</td>\n",
       "      <td>QQLYTFPVT</td>\n",
       "      <td>https://www.rcsb.org/structure/8IX3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hui Sun et al., 2023 (https://journals.asm.org...</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NIV-10</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Delta;SARS-CoV2_Omicron...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Delta;SARS-CoV2_Omicron...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>Computational engineering from another binder</td>\n",
       "      <td>EVQLVESGGGLVQPGGSLRLSCAASGFTFYSYWMTWVRQAPGKGLE...</td>\n",
       "      <td>QSALTQPASVSGSPGQSITISCTGTSSDVGGYNFVSWYRQYPGKAP...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGLJ2 (Human)</td>\n",
       "      <td>ARDGGYNILTAYYHAPSYW</td>\n",
       "      <td>SSYTGRSPYVV</td>\n",
       "      <td>https://www.rcsb.org/structure/8HES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Saya Moriyama et al., 2023 (https://www.nature...</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NIV-11</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Delta;SARS-CoV2_Omicron...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Delta;SARS-CoV2_Omicron...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>Computational engineering from another binder</td>\n",
       "      <td>QLVQSGPEVKKPGTSVKVSCKASGFTFYYSAVQWVRQARGQRLEWL...</td>\n",
       "      <td>EIVLTQSPGTLSLSPGDRAILSCRASQTVNSNYLAWYQQKPGQAPR...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ1 (Human)</td>\n",
       "      <td>AAPNCTGGSCYDGFNL</td>\n",
       "      <td>QQYGSSPWL</td>\n",
       "      <td>https://www.rcsb.org/structure/8HGL;https://ww...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Saya Moriyama et al., 2023 (https://www.nature...</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>P4J15</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_Omicron-XBB1;SARS-CoV2_Omicron-XBB1....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_Omicron-XBB1;SARS-CoV2_Omicron-XBB1....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>B-cells; SARS-CoV2 Human Patient</td>\n",
       "      <td>QVQIQQWGAGLLKPSETLSLTCAVYDESFSDYFWTWIRQSPGMGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTVTCQASQGITNYVNWYQQKPGKAPKL...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ5 (Human)</td>\n",
       "      <td>ARDSPLGSIIKRGDY</td>\n",
       "      <td>QQFDHLPPT</td>\n",
       "      <td>https://www.rcsb.org/structure/8PQ2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Craig Fenwick et al., 2023 (https://www.journa...</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>Feb 8, 2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12913</th>\n",
       "      <td>BD55-6725</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_O...</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>SARS convalescents</td>\n",
       "      <td>EVQVVESGGGLVKPGGSLRLSCAASGFTFSSYTMNWVRQAPGKGLE...</td>\n",
       "      <td>EIVLAQSPATLSLSPGERAILSCRASQSVNTYLAWYQQKPGQAPRL...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ4 (Human)</td>\n",
       "      <td>ARDRFGEFIYPGALDI</td>\n",
       "      <td>QQRSNWPLT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yunlong Cao et al., 2022 (https://www.biorxiv....</td>\n",
       "      <td>Oct 3, 2022</td>\n",
       "      <td>Dec 31, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12914</th>\n",
       "      <td>BD55-6726</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_O...</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>SARS convalescents</td>\n",
       "      <td>QMQLVQSGPEVKRPGTSVKVSCEASGFTFSSSAILWVRQPRGQRLE...</td>\n",
       "      <td>EIVLRQSPATVSLSPGERATLSCRASQSVHNYLAWFQQKPGQAPRL...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ1 (Human)</td>\n",
       "      <td>AAMVRGGWWYFDL</td>\n",
       "      <td>QQRSDWPWT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yunlong Cao et al., 2022 (https://www.biorxiv....</td>\n",
       "      <td>Oct 3, 2022</td>\n",
       "      <td>Dec 31, 2022</td>\n",
       "      <td>21st Mar '23: Overwrote gene assignment from s...</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12915</th>\n",
       "      <td>BD55-6727</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Omicron-BA1 (weak)</td>\n",
       "      <td>SARS-CoV2_Omicron-BA2;SARS-CoV2_Omicron-BA2.75...</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>SARS convalescents</td>\n",
       "      <td>QVQLVESGGGLVKPGGSLRLSCAASGFTFSDYYMNWIRQAPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASVGDRVTITCRASQSISNYLNWYQQKPGKAPKL...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ2 (Human)</td>\n",
       "      <td>ASDPGGLSELAAEYFHH</td>\n",
       "      <td>QQSHSTPYT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yunlong Cao et al., 2022 (https://www.biorxiv....</td>\n",
       "      <td>Oct 3, 2022</td>\n",
       "      <td>Dec 31, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12916</th>\n",
       "      <td>BD55-6728</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_O...</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>SARS convalescents</td>\n",
       "      <td>EVQLVESGGGLVQPGGSLRLSCAASGFTFSRFAMHWVRQAPGKGLE...</td>\n",
       "      <td>DIQMTQSPSSLSASLGDRVTITCRASQTISRYLNWYQQKPGKAPKV...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGKJ1 (Human)</td>\n",
       "      <td>ARESTEVTHWYFDL</td>\n",
       "      <td>QQSYVTPEYT</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yunlong Cao et al., 2022 (https://www.biorxiv....</td>\n",
       "      <td>Oct 3, 2022</td>\n",
       "      <td>Dec 31, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12917</th>\n",
       "      <td>BD55-6729</td>\n",
       "      <td>Ab</td>\n",
       "      <td>SARS-CoV2_WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_O...</td>\n",
       "      <td>S; RBD</td>\n",
       "      <td>SARS convalescents</td>\n",
       "      <td>QVQLVQSGAEVKKPGTSMRVSCKASGYTFSTYGIIWVRQAPGQGLE...</td>\n",
       "      <td>SYELTQPPSVSVSPGQTARITCSGDAVATQFLYWYQQKSGQAPVMV...</td>\n",
       "      <td>...</td>\n",
       "      <td>IGLJ2 (Human)</td>\n",
       "      <td>ARQLLFFGDLSGDNGMDV</td>\n",
       "      <td>QSADSRGVV</td>\n",
       "      <td>ND</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yunlong Cao et al., 2022 (https://www.biorxiv....</td>\n",
       "      <td>Oct 3, 2022</td>\n",
       "      <td>Dec 31, 2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Complete</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8105 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Name Ab or Nb                                           Binds to  \\\n",
       "3           8-D9       Ab  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...   \n",
       "4       Sun_1G11       Ab  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...   \n",
       "11        NIV-10       Ab  SARS-CoV2_WT;SARS-CoV2_Delta;SARS-CoV2_Omicron...   \n",
       "12        NIV-11       Ab  SARS-CoV2_WT;SARS-CoV2_Delta;SARS-CoV2_Omicron...   \n",
       "22         P4J15       Ab  SARS-CoV2_Omicron-XBB1;SARS-CoV2_Omicron-XBB1....   \n",
       "...          ...      ...                                                ...   \n",
       "12913  BD55-6725       Ab                                       SARS-CoV2_WT   \n",
       "12914  BD55-6726       Ab                                       SARS-CoV2_WT   \n",
       "12915  BD55-6727       Ab    SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_WT   \n",
       "12916  BD55-6728       Ab                                       SARS-CoV2_WT   \n",
       "12917  BD55-6729       Ab                                       SARS-CoV2_WT   \n",
       "\n",
       "      Doesn't Bind to                                    Neutralising Vs  \\\n",
       "3                 NaN  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...   \n",
       "4                 NaN  SARS-CoV2_WT;SARS-CoV2_Alpha;SARS-CoV2_Beta;SA...   \n",
       "11                NaN  SARS-CoV2_WT;SARS-CoV2_Delta;SARS-CoV2_Omicron...   \n",
       "12                NaN  SARS-CoV2_WT;SARS-CoV2_Delta;SARS-CoV2_Omicron...   \n",
       "22                NaN  SARS-CoV2_Omicron-XBB1;SARS-CoV2_Omicron-XBB1....   \n",
       "...               ...                                                ...   \n",
       "12913             NaN                                                NaN   \n",
       "12914             NaN                                                NaN   \n",
       "12915             NaN          SARS-CoV2_WT;SARS-CoV2_Omicron-BA1 (weak)   \n",
       "12916             NaN                                                NaN   \n",
       "12917             NaN                                                NaN   \n",
       "\n",
       "                                     Not Neutralising Vs Protein + Epitope  \\\n",
       "3                                                    NaN            S; RBD   \n",
       "4      SARS-CoV2_Omicron-BQ1;SARS-CoV2_Omicron-BQ1.1;...            S; RBD   \n",
       "11                                                   NaN            S; RBD   \n",
       "12                                                   NaN            S; RBD   \n",
       "22                                                   NaN            S; RBD   \n",
       "...                                                  ...               ...   \n",
       "12913  SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_O...            S; RBD   \n",
       "12914  SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_O...            S; RBD   \n",
       "12915  SARS-CoV2_Omicron-BA2;SARS-CoV2_Omicron-BA2.75...            S; RBD   \n",
       "12916  SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_O...            S; RBD   \n",
       "12917  SARS-CoV2_WT;SARS-CoV2_Omicron-BA1;SARS-CoV2_O...            S; RBD   \n",
       "\n",
       "                                               Origin  \\\n",
       "3      B-cells; SARS-CoV2 Human Vaccinee (BBIBP-CorV)   \n",
       "4                    B-cells; SARS-CoV2 Human Patient   \n",
       "11      Computational engineering from another binder   \n",
       "12      Computational engineering from another binder   \n",
       "22                   B-cells; SARS-CoV2 Human Patient   \n",
       "...                                               ...   \n",
       "12913                              SARS convalescents   \n",
       "12914                              SARS convalescents   \n",
       "12915                              SARS convalescents   \n",
       "12916                              SARS convalescents   \n",
       "12917                              SARS convalescents   \n",
       "\n",
       "                                                 VHorVHH  \\\n",
       "3      VQLVESGGGLVQPGGSLRLSCAASGLTVSSNYMNWVRQAPGKGLEW...   \n",
       "4      EVQLVESGGGLVQPGRSLRLSCAASGFKFDDYAMHWVRQAPGKGLE...   \n",
       "11     EVQLVESGGGLVQPGGSLRLSCAASGFTFYSYWMTWVRQAPGKGLE...   \n",
       "12     QLVQSGPEVKKPGTSVKVSCKASGFTFYYSAVQWVRQARGQRLEWL...   \n",
       "22     QVQIQQWGAGLLKPSETLSLTCAVYDESFSDYFWTWIRQSPGMGLE...   \n",
       "...                                                  ...   \n",
       "12913  EVQVVESGGGLVKPGGSLRLSCAASGFTFSSYTMNWVRQAPGKGLE...   \n",
       "12914  QMQLVQSGPEVKRPGTSVKVSCEASGFTFSSSAILWVRQPRGQRLE...   \n",
       "12915  QVQLVESGGGLVKPGGSLRLSCAASGFTFSDYYMNWIRQAPGKGLE...   \n",
       "12916  EVQLVESGGGLVQPGGSLRLSCAASGFTFSRFAMHWVRQAPGKGLE...   \n",
       "12917  QVQLVQSGAEVKKPGTSMRVSCKASGYTFSTYGIIWVRQAPGQGLE...   \n",
       "\n",
       "                                                      VL  ...   Light J Gene  \\\n",
       "3      DIQMTQSPSFLSASVGDRVTITCRASQGISSYLAWYQQKPGKAPKL...  ...  IGKJ3 (Human)   \n",
       "4      DIQLTQSPSFLSASVGDRVTITCRASQGIDKYLAWYQQKPGQAPKV...  ...  IGKJ3 (Human)   \n",
       "11     QSALTQPASVSGSPGQSITISCTGTSSDVGGYNFVSWYRQYPGKAP...  ...  IGLJ2 (Human)   \n",
       "12     EIVLTQSPGTLSLSPGDRAILSCRASQTVNSNYLAWYQQKPGQAPR...  ...  IGKJ1 (Human)   \n",
       "22     DIQMTQSPSSLSASVGDRVTVTCQASQGITNYVNWYQQKPGKAPKL...  ...  IGKJ5 (Human)   \n",
       "...                                                  ...  ...            ...   \n",
       "12913  EIVLAQSPATLSLSPGERAILSCRASQSVNTYLAWYQQKPGQAPRL...  ...  IGKJ4 (Human)   \n",
       "12914  EIVLRQSPATVSLSPGERATLSCRASQSVHNYLAWFQQKPGQAPRL...  ...  IGKJ1 (Human)   \n",
       "12915  DIQMTQSPSSLSASVGDRVTITCRASQSISNYLNWYQQKPGKAPKL...  ...  IGKJ2 (Human)   \n",
       "12916  DIQMTQSPSSLSASLGDRVTITCRASQTISRYLNWYQQKPGKAPKV...  ...  IGKJ1 (Human)   \n",
       "12917  SYELTQPPSVSVSPGQTARITCSGDAVATQFLYWYQQKSGQAPVMV...  ...  IGLJ2 (Human)   \n",
       "\n",
       "                     CDRH3        CDRL3  \\\n",
       "3              ARDHSGHALDI  QHLNSYPSMYT   \n",
       "4       VKDSNYDSSGYLINNFDY    QQLYTFPVT   \n",
       "11     ARDGGYNILTAYYHAPSYW  SSYTGRSPYVV   \n",
       "12        AAPNCTGGSCYDGFNL    QQYGSSPWL   \n",
       "22         ARDSPLGSIIKRGDY    QQFDHLPPT   \n",
       "...                    ...          ...   \n",
       "12913     ARDRFGEFIYPGALDI    QQRSNWPLT   \n",
       "12914        AAMVRGGWWYFDL    QQRSDWPWT   \n",
       "12915    ASDPGGLSELAAEYFHH    QQSHSTPYT   \n",
       "12916       ARESTEVTHWYFDL   QQSYVTPEYT   \n",
       "12917   ARQLLFFGDLSGDNGMDV    QSADSRGVV   \n",
       "\n",
       "                                              Structures  \\\n",
       "3      https://www.rcsb.org/structure/8J1T;https://ww...   \n",
       "4                    https://www.rcsb.org/structure/8IX3   \n",
       "11                   https://www.rcsb.org/structure/8HES   \n",
       "12     https://www.rcsb.org/structure/8HGL;https://ww...   \n",
       "22                   https://www.rcsb.org/structure/8PQ2   \n",
       "...                                                  ...   \n",
       "12913                                                 ND   \n",
       "12914                                                 ND   \n",
       "12915                                                 ND   \n",
       "12916                                                 ND   \n",
       "12917                                                 ND   \n",
       "\n",
       "      ABB Homology Model (if no structure)  \\\n",
       "3                                      NaN   \n",
       "4                                      NaN   \n",
       "11                                     NaN   \n",
       "12                                     NaN   \n",
       "22                                     NaN   \n",
       "...                                    ...   \n",
       "12913                                  NaN   \n",
       "12914                                  NaN   \n",
       "12915                                  NaN   \n",
       "12916                                  NaN   \n",
       "12917                                  NaN   \n",
       "\n",
       "                                                 Sources   Date Added  \\\n",
       "3      Wanbo Tai et al., 2023 (https://www.nature.com...  Feb 8, 2024   \n",
       "4      Hui Sun et al., 2023 (https://journals.asm.org...  Feb 8, 2024   \n",
       "11     Saya Moriyama et al., 2023 (https://www.nature...  Feb 8, 2024   \n",
       "12     Saya Moriyama et al., 2023 (https://www.nature...  Feb 8, 2024   \n",
       "22     Craig Fenwick et al., 2023 (https://www.journa...  Feb 8, 2024   \n",
       "...                                                  ...          ...   \n",
       "12913  Yunlong Cao et al., 2022 (https://www.biorxiv....  Oct 3, 2022   \n",
       "12914  Yunlong Cao et al., 2022 (https://www.biorxiv....  Oct 3, 2022   \n",
       "12915  Yunlong Cao et al., 2022 (https://www.biorxiv....  Oct 3, 2022   \n",
       "12916  Yunlong Cao et al., 2022 (https://www.biorxiv....  Oct 3, 2022   \n",
       "12917  Yunlong Cao et al., 2022 (https://www.biorxiv....  Oct 3, 2022   \n",
       "\n",
       "       Last Updated                                 Update Description  \\\n",
       "3       Feb 8, 2024                                                NaN   \n",
       "4       Feb 8, 2024                                                NaN   \n",
       "11      Feb 8, 2024                                                NaN   \n",
       "12      Feb 8, 2024                                                NaN   \n",
       "22      Feb 8, 2024                                                NaN   \n",
       "...             ...                                                ...   \n",
       "12913  Dec 31, 2022                                                NaN   \n",
       "12914  Dec 31, 2022  21st Mar '23: Overwrote gene assignment from s...   \n",
       "12915  Dec 31, 2022                                                NaN   \n",
       "12916  Dec 31, 2022                                                NaN   \n",
       "12917  Dec 31, 2022                                                NaN   \n",
       "\n",
       "      Notes/Following Up?  \n",
       "3                Complete  \n",
       "4                Complete  \n",
       "11               Complete  \n",
       "12               Complete  \n",
       "22               Complete  \n",
       "...                   ...  \n",
       "12913            Complete  \n",
       "12914            Complete  \n",
       "12915            Complete  \n",
       "12916            Complete  \n",
       "12917            Complete  \n",
       "\n",
       "[8105 rows x 23 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_epitopes = ['S; RBD', 'S; iso-RBD', 'S: RBD', 'S; RBD/non-RBD', 'S; probably RBD (implied by clustering)']\n",
    "epitope_mask = df['Protein + Epitope'].isin(valid_epitopes)\n",
    "\n",
    "vh_valid_mask = df['VHorVHH'].str.len() > 50\n",
    "\n",
    "filtered = df[epitope_mask & vh_valid_mask].copy()\n",
    "filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "088d9f76-5a95-4be2-a20f-aae71606e1c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences = filtered[\"VHorVHH\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8fe6cc7-b6f6-4d80-b966-ae57a86aae52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'VQLVESGGGLVQPGGSLRLSCAASGLTVSSNYMNWVRQAPGKGLEWVSVFYPGGSTFYADSVRGRFTISRDNSKNTLYLQMNSLRAEDTAVYYCARDHSGHALDIWGQGTMVTVS'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "87d6cbd8-19ff-4d71-88f5-f647c446bc9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Limiting hmmer search to species ['human', 'mouse'] was requested but hits did not achieve a high enough bitscore. Reverting to using any species\n",
      "Limiting hmmer search to species ['human', 'mouse'] was requested but hits did not achieve a high enough bitscore. Reverting to using any species\n"
     ]
    }
   ],
   "source": [
    "from protein.helpers import extract_regions\n",
    "parsed = extract_regions(\n",
    "    sequences,\n",
    "    save_txt=\"data/full_vhs.txt\",\n",
    "    save_csv=\"data/vh_regions.csv\",\n",
    "    # save_metadata=\"data/vh_metadata.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4a2627c7-58a6-4b51-bfee-f8ab21d5ffe1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'FR1': 'EVQLVESGGGLVQPGRSLRLSCAAS',\n",
       " 'CDR1': 'GFKFDDYA',\n",
       " 'FR2': 'MHWVRQAPGKGLEWVSG',\n",
       " 'CDR2': 'TSWNSGTT',\n",
       " 'FR3': 'GYADSVRGRFTISRDNAKKSLYLQMNSLGVEDTAFYYC',\n",
       " 'CDR3': 'VKDSNYDSSGYLINNFDY',\n",
       " 'FR4': 'WGQGILVTVSS',\n",
       " 'full': 'EVQLVESGGGLVQPGRSLRLSCAASGFKFDDYAMHWVRQAPGKGLEWVSGTSWNSGTTGYADSVRGRFTISRDNAKKSLYLQMNSLGVEDTAFYYCVKDSNYDSSGYLINNFDYWGQGILVTVSS',\n",
       " 'missing_regions': []}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parsed[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bb39b4cc-d667-4631-900e-da1d5b8411bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train / Val / Test split\n",
    "train_seqs, other_seqs = train_test_split(sequences, test_size=0.1, random_state=970713)\n",
    "test_seqs, val_seqs = train_test_split(other_seqs, test_size=0.5, random_state=970713)\n",
    "\n",
    "# Save as FASTA for CD-HIT and TFRecord processing\n",
    "def save_fasta(sequences, filename):\n",
    "    with open(filename, \"w\") as f:\n",
    "        for i, seq in enumerate(sequences):\n",
    "            f.write(f\">seq{i}\\n{seq}\\n\")\n",
    "\n",
    "save_fasta(train_seqs, \"data/zz/train.fasta\")\n",
    "save_fasta(val_seqs, \"data/zz/val.fasta\")\n",
    "save_fasta(test_seqs, \"data/zz/test.fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "09e8eeb9-2605-4f1a-a382-cc5926dc974a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "def fasta_to_dataframe(fasta_path):\n",
    "    records = []\n",
    "    for record in SeqIO.parse(fasta_path, \"fasta\"):\n",
    "        records.append({\"id\": record.id, \"sequence\": str(record.seq)})\n",
    "    return pd.DataFrame(records)\n",
    "\n",
    "df = fasta_to_dataframe(\"data/zz/train.fasta\")\n",
    "df.to_csv(\"data/zz/train.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e8f84fb-c138-4e06-9d40-36dd62eb1196",
   "metadata": {},
   "outputs": [],
   "source": [
    "def aa_to_int(sequence, padding_length=None):\n",
    "    \"\"\"\n",
    "    Converts an amino acid sequence to a list of integers (1-based index).\n",
    "    Optionally pads the sequence to a fixed length with 0s.\n",
    "    \"\"\"\n",
    "    aa_map = {\n",
    "        'A': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5,\n",
    "        'G': 6, 'H': 7, 'I': 8, 'K': 9, 'L': 10,\n",
    "        'M': 11, 'N': 12, 'P': 13, 'Q': 14, 'R': 15,\n",
    "        'S': 16, 'T': 17, 'V': 18, 'W': 19, 'Y': 20\n",
    "    }\n",
    "\n",
    "    encoded = [aa_map.get(aa, 0) for aa in sequence.upper()]  # unknowns → 0\n",
    "\n",
    "    if padding_length:\n",
    "        if len(encoded) > padding_length:\n",
    "            encoded = encoded[:padding_length]\n",
    "        else:\n",
    "            encoded += [0] * (padding_length - len(encoded))\n",
    "\n",
    "    return encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c04176f-845e-42ca-bfb7-d0a7eaf8f0d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from collections import defaultdict\n",
    "\n",
    "def parse_clstr_file(clstr_path):\n",
    "    \"\"\"\n",
    "    Parses a .clstr file and returns a list of clusters,\n",
    "    each cluster is a list of sequence names.\n",
    "    \"\"\"\n",
    "    clusters = []\n",
    "    with open(clstr_path, \"r\") as f:\n",
    "        current = []\n",
    "        for line in f:\n",
    "            if line.startswith(\">Cluster\"):\n",
    "                if current:\n",
    "                    clusters.append(current)\n",
    "                current = []\n",
    "            else:\n",
    "                seq_id = line.split(\">\")[1].split(\"...\")[0]\n",
    "                current.append(seq_id)\n",
    "        if current:\n",
    "            clusters.append(current)\n",
    "    return clusters\n",
    "\n",
    "def compute_upsampling(clusters, target_size=32, max_factor=50):\n",
    "    \"\"\"\n",
    "    Assign upsampling factors to each sequence based on target cluster size.\n",
    "    \n",
    "    Args:\n",
    "        clusters: list of clusters (each cluster is a list of sequence IDs)\n",
    "        target_size: the target cluster size to normalize upsampling\n",
    "        max_factor: optional cap to avoid runaway inflation of upsampling\n",
    "\n",
    "    Returns:\n",
    "        upsampling_map: dict mapping sequence ID → upsampling factor\n",
    "    \"\"\"\n",
    "    upsampling_map = {}\n",
    "    upsampling_factors = []\n",
    "\n",
    "    for cluster in clusters:\n",
    "        cluster_size = len(cluster)\n",
    "        factor = max(1, target_size // cluster_size)\n",
    "        factor = min(factor, max_factor)\n",
    "\n",
    "        for seq_id in cluster:\n",
    "            upsampling_map[seq_id] = factor\n",
    "            upsampling_factors.append(factor)\n",
    "\n",
    "    # Logging\n",
    "    import numpy as np\n",
    "    print(\"Upsampling factor stats:\")\n",
    "    print(f\"  Min: {min(upsampling_factors)}\")\n",
    "    print(f\"  Max: {max(upsampling_factors)}\")\n",
    "    print(f\"  Mean: {np.mean(upsampling_factors):.2f}\")\n",
    "    print(f\"  Total synthetic samples after upsampling: {sum(upsampling_factors)} (original: {sum(len(c) for c in clusters)})\")\n",
    "\n",
    "    return upsampling_map\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "57009dde-e558-493a-a097-6819aa432999",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import tensorflow as tf\n",
    "from collections import defaultdict\n",
    "\n",
    "def assign_bin(factor):\n",
    "    if factor <= 1:\n",
    "        return 'bin_1'\n",
    "    elif factor <= 2:\n",
    "        return 'bin_2'\n",
    "    elif factor <= 4:\n",
    "        return 'bin_4'\n",
    "    elif factor <= 8:\n",
    "        return 'bin_8'\n",
    "    elif factor <= 16:\n",
    "        return 'bin_16'\n",
    "    else:\n",
    "        return 'bin_32'\n",
    "\n",
    "\n",
    "def write_clustered_tfrecords(csv_file, upsampling_map, output_dir):\n",
    "    df = pd.read_csv(csv_file)\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    # Group sequence IDs into bins\n",
    "    bin_groups = defaultdict(list)\n",
    "    for seq_id, factor in upsampling_map.items():\n",
    "        bin_name = assign_bin(factor)\n",
    "        bin_groups[bin_name].append((seq_id, factor))\n",
    "\n",
    "    # Write a TFRecord file for each bin\n",
    "    for bin_name, entries in bin_groups.items():\n",
    "        filename = os.path.join(output_dir, f\"{bin_name}.tfrecords\")\n",
    "        with tf.io.TFRecordWriter(filename) as writer:\n",
    "            for seq_id, factor in entries:\n",
    "                seq_row = df[df['id'] == seq_id]\n",
    "                if seq_row.empty:\n",
    "                    continue\n",
    "                seq = seq_row['sequence'].values[0]\n",
    "                encoded = aa_to_int(seq)\n",
    "\n",
    "                for _ in range(factor):\n",
    "                    feature = {\n",
    "                        \"label\": tf.train.Feature(int64_list=tf.train.Int64List(value=[0])),  # Dummy label\n",
    "                        \"sequence\": tf.train.Feature(int64_list=tf.train.Int64List(value=encoded))\n",
    "                    }\n",
    "                    example = tf.train.Example(features=tf.train.Features(feature=feature))\n",
    "                    writer.write(example.SerializeToString())\n",
    "        print(f\"{bin_name}: {len(entries)} sequences\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eed1260a-c21c-41d8-9046-a796afd5f0a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def sample_real_sequences_proportional(clusters, total_target=500, max_per_cluster=20, seed=42):\n",
    "    \"\"\"\n",
    "    Samples real sequence IDs proportionally by cluster size, capped by max_per_cluster.\n",
    "    \n",
    "    Args:\n",
    "        clusters (list of list of str): List of clusters, each a list of sequence IDs.\n",
    "        total_target (int): Approximate total number of sequences to sample.\n",
    "        max_per_cluster (int): Cap on how many sequences to sample per cluster.\n",
    "        seed (int): Random seed for reproducibility.\n",
    "    \n",
    "    Returns:\n",
    "        sampled_ids (list of str): List of sampled sequence IDs.\n",
    "    \"\"\"\n",
    "    random.seed(seed)\n",
    "    total_seqs = sum(len(c) for c in clusters)\n",
    "    sampled_ids = []\n",
    "\n",
    "    for cluster in clusters:\n",
    "        proportion = len(cluster) / total_seqs\n",
    "        n = min(max_per_cluster, max(1, int(proportion * total_target)))\n",
    "        sampled_ids.extend(random.sample(cluster, min(n, len(cluster))))\n",
    "\n",
    "    # Adjust if we overshot or undershot the target\n",
    "    if len(sampled_ids) > total_target:\n",
    "        sampled_ids = random.sample(sampled_ids, total_target)\n",
    "    elif len(sampled_ids) < total_target:\n",
    "        # Top up randomly (optional)\n",
    "        remaining = total_target - len(sampled_ids)\n",
    "        flat = [sid for cluster in clusters for sid in cluster if sid not in sampled_ids]\n",
    "        sampled_ids.extend(random.sample(flat, min(remaining, len(flat))))\n",
    "\n",
    "    return sampled_ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "18a20a30-965e-478e-8f2c-eaa0285818f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upsampling factor stats:\n",
      "  Min: 1\n",
      "  Max: 32\n",
      "  Mean: 5.72\n",
      "  Total synthetic samples after upsampling: 40594 (original: 7097)\n"
     ]
    }
   ],
   "source": [
    "clusters = parse_clstr_file(\"data/zz/clustered_train_80.fasta.clstr\")\n",
    "upsample_map = compute_upsampling(clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47229345-15a0-40f6-bdbf-778410747306",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_seq_ids = sample_real_sequences_proportional(clusters, total_target=500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "eb929c61-6462-4558-8267-1f3ed54189be",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"data/zz/train.csv\")\n",
    "sample_seqs = df[df['id'].isin(sampled_seq_ids)].reset_index(drop=True) # For comparing embeddings vs generated "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "71d02cb6-8f37-48c5-8713-eda2e973ddcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_seqs.to_csv(\"data/zz/sample_seqs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c43e6634-9ebb-4f5b-a2cb-7bf119a6a5fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total clusters: 1180\n",
      "Max: 518\n",
      "Median: 1.0\n",
      "95th percentile: 16.0\n"
     ]
    }
   ],
   "source": [
    "cluster_sizes = [len(c) for c in clusters]\n",
    "import numpy as np\n",
    "\n",
    "print(\"Total clusters:\", len(cluster_sizes))\n",
    "print(\"Max:\", max(cluster_sizes))\n",
    "print(\"Median:\", np.median(cluster_sizes))\n",
    "print(\"95th percentile:\", np.percentile(cluster_sizes, 95))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1bdcfa43-55c7-4055-9260-0c0e441b8e1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bin_1: 5045 sequences\n",
      "bin_16: 414 sequences\n",
      "bin_8: 319 sequences\n",
      "bin_2: 324 sequences\n",
      "bin_4: 167 sequences\n",
      "bin_32: 828 sequences\n"
     ]
    }
   ],
   "source": [
    "write_clustered_tfrecords(\"data/zz/train.csv\", upsample_map, \"data/zz/train\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b473f39-04e9-4fad-9176-5de657c90aca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tf2_16",
   "language": "python",
   "name": "tf2_16"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.22"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
